## Import data

In [1]:
import pandas as pd
import numpy as np
df = pd.read_csv('Combined_Flights_2022.csv')
print(f"Original shape: {df.shape}")
df.head()

Original shape: (4078318, 61)


,FlightDate,Airline,Origin,Dest,Cancelled,Diverted,CRSDepTime,DepTime,DepDelayMinutes,DepDelay,...,WheelsOff,WheelsOn,TaxiIn,CRSArrTime,ArrDelay,ArrDel15,ArrivalDelayGroups,ArrTimeBlk,DistanceGroup,DivAirportLandings
0,2022-04-04,"Commutair Aka Champlain Enterprises, Inc.",GJT,DEN,False,False,1133,1123.0,0.0,-10.0,...,1140.0,1220.0,8.0,1245,-17.0,0.0,-2.0,1200-1259,1,0
1,2022-04-04,"Commutair Aka Champlain Enterprises, Inc.",HRL,IAH,False,False,732,728.0,0.0,-4.0,...,744.0,839.0,9.0,849,-1.0,0.0,-1.0,0800-0859,2,0
2,2022-04-04,"Commutair Aka Champlain Enterprises, Inc.",DRO,DEN,False,False,1529,1514.0,0.0,-15.0,...,1535.0,1622.0,14.0,1639,-3.0,0.0,-1.0,1600-1659,2,0
3,2022-04-04,"Commutair Aka Champlain Enterprises, Inc.",IAH,GPT,False,False,1435,1430.0,0.0,-5.0,...,1446.0,1543.0,4.0,1605,-18.0,0.0,-2.0,1600-1659,2,0
4,2022-04-04,"Commutair Aka Champlain Enterprises, Inc.",DRO,DEN,False,False,1135,1135.0,0.0,0.0,...,1154.0,1243.0,8.0,1245,6.0,0.0,0.0,1200-1259,2,0


## EDA

In [2]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4078318 entries, 0 to 4078317
Data columns (total 61 columns):
 #   Column                                   Dtype  
---  ------                                   -----  
 0   FlightDate                               object 
 1   Airline                                  object 
 2   Origin                                   object 
 3   Dest                                     object 
 4   Cancelled                                bool   
 5   Diverted                                 bool   
 6   CRSDepTime                               int64  
 7   DepTime                                  float64
 8   DepDelayMinutes                          float64
 9   DepDelay                                 float64
 10  ArrTime                                  float64
 11  ArrDelayMinutes                          float64
 12  AirTime                                  float64
 13  CRSElapsedTime                           float64
 14  ActualElapsedTime 

In [3]:
df.describe()

,CRSDepTime,DepTime,DepDelayMinutes,DepDelay,ArrTime,ArrDelayMinutes,AirTime,CRSElapsedTime,ActualElapsedTime,Distance,...,TaxiOut,WheelsOff,WheelsOn,TaxiIn,CRSArrTime,ArrDelay,ArrDel15,ArrivalDelayGroups,DistanceGroup,DivAirportLandings
count,4.078318e+06,3.957885e+06,3.957823e+06,3.957823e+06,3.954079e+06,3.944916e+06,3.944916e+06,4.078318e+06,3.944916e+06,4.078318e+06,...,3.955652e+06,3.955652e+06,3.954076e+06,3.954076e+06,4.078318e+06,3.944916e+06,3.944916e+06,3.944916e+06,4.078318e+06,4.078318e+06
mean,1.329587e+03,1.334374e+03,1.601494e+01,1.309049e+01,1.457886e+03,1.578307e+01,1.110075e+02,1.413211e+02,1.358624e+02,7.978657e+02,...,1.697375e+01,1.356576e+03,1.455073e+03,7.894387e+00,1.486058e+03,7.528486e+00,2.164715e-01,-6.256103e-02,3.663516e+00,3.685098e-03
std,4.904801e+02,5.056219e+02,5.231498e+01,5.332016e+01,5.431841e+02,5.198424e+01,6.996246e+01,7.179635e+01,7.185501e+01,5.914742e+02,...,9.495407e+00,5.075580e+02,5.378428e+02,6.663118e+00,5.185078e+02,5.524625e+01,4.118393e-01,2.487442e+00,2.320848e+00,1.141331e-01
min,1.000000e+00,1.000000e+00,0.000000e+00,-7.800000e+01,1.000000e+00,0.000000e+00,8.000000e+00,-4.800000e+01,1.400000e+01,3.100000e+01,...,1.000000e+00,1.000000e+00,1.000000e+00,1.000000e+00,1.000000e+00,-1.000000e+02,0.000000e+00,-2.000000e+00,1.000000e+00,0.000000e+00
25%,9.140000e+02,9.170000e+02,0.000000e+00,-5.000000e+00,1.046000e+03,0.000000e+00,6.000000e+01,8.900000e+01,8.300000e+01,3.680000e+02,...,1.100000e+01,9.320000e+02,1.044000e+03,4.000000e+00,1.103000e+03,-1.400000e+01,0.000000e+00,-1.000000e+00,2.000000e+00,0.000000e+00
50%,1.320000e+03,1.325000e+03,0.000000e+00,-2.000000e+00,1.500000e+03,0.000000e+00,9.400000e+01,1.240000e+02,1.190000e+02,6.430000e+02,...,1.500000e+01,1.338000e+03,1.456000e+03,6.000000e+00,1.513000e+03,-5.000000e+00,0.000000e+00,-1.000000e+00,3.000000e+00,0.000000e+00
75%,1.735000e+03,1.744000e+03,1.100000e+01,1.100000e+01,1.914000e+03,1.000000e+01,1.410000e+02,1.710000e+02,1.670000e+02,1.035000e+03,...,1.900000e+01,1.758000e+03,1.909000e+03,9.000000e+00,1.920000e+03,1.000000e+01,0.000000e+00,0.000000e+00,5.000000e+00,0.000000e+00
max,2.359000e+03,2.400000e+03,7.223000e+03,7.223000e+03,2.400000e+03,7.232000e+03,7.270000e+02,6.900000e+02,7.640000e+02,5.095000e+03,...,2.210000e+02,2.400000e+03,2.400000e+03,2.900000e+02,2.359000e+03,7.232000e+03,1.000000e+00,1.200000e+01,1.100000e+01,9.000000e+00


## Check dublicates

In [4]:
print(f"duplicates in data: {df.duplicated().sum()}")

duplicates in data: 0


* There is no dublicates in data

## Check nulls & handling it

In [5]:
df.isnull().sum()

FlightDate                 0
Airline                    0
Origin                     0
Dest                       0
Cancelled                  0
                       ...  
ArrDel15              133402
ArrivalDelayGroups    133402
ArrTimeBlk                 0
DistanceGroup              0
DivAirportLandings         0
Length: 61, dtype: int64

* في ترتيب واضح ف نالز فهنبدأ نستكشف الداتا من غير الرحلات الملغيه والمحوله

In [6]:

cancelled_df = df[df['Cancelled'] == True].copy()


diverted_df = df[df['Diverted'] == True].copy()

flights_df = df[(df['Cancelled'] == False) & (df['Diverted'] == False)].copy()


print(flights_df[['DepTime','ArrTime','AirTime']].isnull().sum())

DepTime    0
ArrTime    0
AirTime    0
dtype: int64


### 3. Null Value Handling
Nulls were **not random** — they followed a logical pattern:

| Column Group | Null Count | Reason |
|---|---|---|
| DepTime, DepDelay... | ~120K | Flight was **Cancelled** — never departed |
| ArrTime, ArrDelay... | ~133K | Flight was **Cancelled or Diverted** — never arrived at destination |
| Tail_Number | ~26K | Data entry issue |

**Strategy — fill with `-1` (not drop):**
- `Cancelled` flights → all delay columns = `-1`
- `Diverted` flights → arrival columns only = `-1`
- `Tail_Number` → filled with `'UNKNOWN'`

> `-1` was chosen over `0` to distinguish *"no data"* from *"zero delay"*

In [7]:
# Numeric col el nulls
delay_cols = ['DepTime', 'DepDelay', 'DepDelayMinutes', 'ArrTime', 'ArrDelay', 
              'ArrDelayMinutes', 'AirTime', 'ActualElapsedTime', 'ArrDel15',
              'DepDel15', 'TaxiOut', 'TaxiIn', 'WheelsOff', 'WheelsOn',
              'DepartureDelayGroups', 'ArrivalDelayGroups']

#  لو اتكنسلت هنعوض ب -1
cancelled_mask = df['Cancelled'] == True
df.loc[cancelled_mask, delay_cols] = df.loc[cancelled_mask, delay_cols].fillna(-1)


diverted_mask = df['Diverted'] == True
arr_cols = ['ArrTime', 'ArrDelay', 'ArrDelayMinutes', 'ArrDel15', 
            'ArrivalDelayGroups', 'TaxiIn', 'WheelsOn', 'ActualElapsedTime', 'AirTime']

df.loc[diverted_mask, arr_cols] = df.loc[diverted_mask, arr_cols].fillna(-1)

# 3. Tail_Number للكل
df['Tail_Number'] = df['Tail_Number'].fillna('UNKNOWN')


print(df.isnull().sum()[df.isnull().sum() > 0])


Series([], dtype: int64)


In [8]:
df['DepDelay']

0          -10.0
1           -4.0
2          -15.0
3           -5.0
4            0.0
           ...  
4078313     25.0
4078314     44.0
4078315    378.0
4078316    113.0
4078317     -6.0
Name: DepDelay, Length: 4078318, dtype: float64

### 4. Flight Status Column
Instead of splitting into multiple DataFrames, we added a single `FlightStatus` column:
- `Completed` — flight departed and arrived normally
- `Cancelled` — flight never departed
- `Diverted` — flight landed at a different airport

This keeps the dataset unified and makes filtering easy in both Python and Power BI.

In [13]:
df['FlightStatus'] = 'Completed'
df.loc[df['Cancelled'] == True, 'FlightStatus'] = 'Cancelled'
df.loc[df['Diverted'] == True, 'FlightStatus'] = 'Diverted'

completed = df[df['FlightStatus'] == 'Completed']

### 5. Data Type Fixes

| Column | Before | After | Reason |
|---|---|---|---|
| `FlightDate` | object | datetime64 | Enable time-based analysis |
| `DepTime`, `ArrTime`... | HHMM float | minutes from midnight | `715` → `435 min` — enables math operations |
| `Airline`, `Origin`... | object | category | Reduces memory significantly |

> Time columns like `DepTime` were stored as HHMM integers (e.g. `715` = 7:15 AM).  
> We converted them to **minutes from midnight** for accurate calculations.

In [14]:
df['DepTime'].unique()

array([1123.,  728., 1514., ...,  437.,  420.,  428.])

In [15]:
def hhmm_to_minutes(val):
    """715.0 → 435 دقيقة من منتصف الليل"""
    if val == -1 or pd.isna(val):
        return -1
    val = int(val)
    hours = val // 100
    minutes = val % 100
    if hours == 24:
        hours = 0
    return hours * 60 + minutes

hhmm_cols = ['CRSDepTime', 'DepTime', 'CRSArrTime', 'ArrTime', 'WheelsOff', 'WheelsOn']

for col in hhmm_cols:
    df[col] = df[col].apply(hhmm_to_minutes)

# تحقق
print(df[['CRSDepTime', 'DepTime', 'ArrTime']].head())
print(df[['CRSDepTime', 'DepTime', 'ArrTime']].describe())

   CRSDepTime  DepTime  ArrTime
0         693      683      748
1         452      448      528
2         929      914      996
3         875      870      947
4         695      695      771
         CRSDepTime       DepTime       ArrTime
count  4.078318e+06  4.078318e+06  4.078318e+06
mean   8.085745e+02  7.886011e+02  8.587756e+02
std    2.954708e+02  3.290224e+02  3.558507e+02
min    1.000000e+00 -1.000000e+00 -1.000000e+00
25%    5.540000e+02  5.330000e+02  6.200000e+02
50%    8.000000e+02  7.890000e+02  8.820000e+02
75%    1.055000e+03  1.056000e+03  1.145000e+03
max    1.439000e+03  1.439000e+03  1.439000e+03


In [16]:
# Datetime
df['FlightDate'] = pd.to_datetime(df['FlightDate'])

# Categories (بتوفر memory كتير مع الداتا الكبيرة دي)
cat_cols = ['Airline', 'Origin', 'Dest', 'FlightStatus',
            'OriginState', 'DestState', 'IATA_Code_Marketing_Airline',
            'IATA_Code_Operating_Airline', 'DepTimeBlk', 'ArrTimeBlk']

df[cat_cols] = df[cat_cols].astype('category')

# memory بعد التحويل
print(f"Memory usage: {df.memory_usage(deep=True).sum() / 1e9:.2f} GB")

Memory usage: 3.26 GB


## 6. Feature Engineering
New columns created to enrich analysis:

| Feature | Description |
|---|---|
| `Flight_Month`, `Flight_Day`, `Flight_DayOfWeek` | Extracted from `FlightDate` |
| `Is_Weekend` | 1 if Saturday or Sunday |
| `MonthName`, `DayName`, `Season` | Human-readable time labels |
| `DepTimeOfDay` | Red-Eye / Morning / Afternoon / Evening |
| `IsDelayed` | 1 if arrival delay > 15 min, -1 if not completed |
| `DelayCategory` | On Time / Minor / Moderate / Severe / Cancelled / Diverted |
| `RouteKey` | `"JFK→LAX"` format for route analysis |
| `DistanceCategory` | Short / Medium / Long / Ultra |
| `ScheduleAdherence` | Difference between scheduled and actual elapsed time |

In [17]:
df['Flight_Month']     = df['FlightDate'].dt.month
df['Flight_Day']       = df['FlightDate'].dt.day
df['Flight_DayOfWeek'] = df['FlightDate'].dt.dayofweek
df['Is_Weekend']       = df['Flight_DayOfWeek'].isin([5, 6]).astype(int)

month_map = {1:'Jan',2:'Feb',3:'Mar',4:'Apr',5:'May',6:'Jun',
             7:'Jul',8:'Aug',9:'Sep',10:'Oct',11:'Nov',12:'Dec'}
df['MonthName'] = df['Flight_Month'].map(month_map).astype('category')

day_map = {0:'Mon',1:'Tue',2:'Wed',3:'Thu',4:'Fri',5:'Sat',6:'Sun'}
df['DayName'] = df['Flight_DayOfWeek'].map(day_map).astype('category')

season_map = {12:'Winter',1:'Winter',2:'Winter',
              3:'Spring',4:'Spring',5:'Spring',
              6:'Summer',7:'Summer',8:'Summer',
              9:'Fall',10:'Fall',11:'Fall'}
df['Season'] = df['Flight_Month'].map(season_map).astype('category')

# ── من وقت الإقلاع — vectorized ──────────
conditions = [
    df['CRSDepTime'] == -1,
    df['CRSDepTime'] < 360,
    df['CRSDepTime'] < 720,
    df['CRSDepTime'] < 1020,
]
choices = ['Unknown', 'Red-Eye', 'Morning', 'Afternoon']
df['DepTimeOfDay'] = np.select(conditions, choices, default='Evening')
df['DepTimeOfDay'] = df['DepTimeOfDay'].astype('category')

# ── من الـ Delay — vectorized ─────────────
df['IsDelayed'] = (df['ArrDelay'] > 15).astype(int)
df.loc[df['FlightStatus'] != 'Completed', 'IsDelayed'] = -1

conditions = [
    df['FlightStatus'] == 'Cancelled',
    df['FlightStatus'] == 'Diverted',
    df['ArrDelay'] <= 0,
    df['ArrDelay'] <= 15,
    df['ArrDelay'] <= 60,
]
choices = ['Cancelled', 'Diverted', 'On Time', 'Minor', 'Moderate']
df['DelayCategory'] = np.select(conditions, choices, default='Severe')
df['DelayCategory'] = df['DelayCategory'].astype('category')

# ── من المطارات ──────────────────────────
df['RouteKey'] = df['Origin'].astype(str) + '→' + df['Dest'].astype(str)
df['RouteKey'] = df['RouteKey'].astype('category')

# ── من المسافة ───────────────────────────
df['DistanceCategory'] = pd.cut(
    df['Distance'],
    bins=[0, 500, 1000, 2000, 6000],
    labels=['Short', 'Medium', 'Long', 'Ultra']
).astype('category')

# ── Efficiency ───────────────────────────
mask = df['FlightStatus'] == 'Completed'
df.loc[mask, 'ScheduleAdherence'] = (
    df.loc[mask, 'CRSElapsedTime'] - df.loc[mask, 'ActualElapsedTime']
)
df['ScheduleAdherence'] = df['ScheduleAdherence'].fillna(-1)

print("Features added ")
print(df[['DelayCategory','DepTimeOfDay','Season','RouteKey','DistanceCategory']].head())
print(f"\nMemory: {df.memory_usage(deep=True).sum() / 1e9:.2f} GB")

Features added 
  DelayCategory DepTimeOfDay  Season RouteKey DistanceCategory
0       On Time      Morning  Spring  GJT→DEN            Short
1       On Time      Morning  Spring  HRL→IAH            Short
2       On Time    Afternoon  Spring  DRO→DEN            Short
3       On Time    Afternoon  Spring  IAH→GPT            Short
4         Minor      Morning  Spring  DRO→DEN            Short

Memory: 3.41 GB


## 7. Dropping Redundant Columns
Removed **20 columns** that were either duplicates or unnecessary:
- Numeric IDs (Airport IDs, Market IDs, WAC codes) — we have names already
- Marketing vs Operating duplicates
- `Cancelled` & `Diverted` booleans — replaced by `FlightStatus`
- Time block columns — replaced by `DepTimeOfDay`

| | Count |
|---|---|
| Before | 70 columns |
| After | **50 columns** |
| Memory saved | ~1 GB |

In [18]:
cols_to_drop = [
    # IDs مش محتاجينها — عندنا الأسماء
    'OriginAirportID', 'OriginAirportSeqID', 'OriginCityMarketID',
    'OriginStateFips', 'OriginWac',
    'DestAirportID', 'DestAirportSeqID', 'DestCityMarketID',
    'DestStateFips', 'DestWac',
    'DOT_ID_Marketing_Airline', 'DOT_ID_Operating_Airline',
     'Year','Quarter','Month',  'DayofMonth' , 'DayOfWeek', 

    # Marketing vs Operating — redundant في معظم الحالات
    'Marketing_Airline_Network', 'Operated_or_Branded_Code_Share_Partners',
    'Flight_Number_Marketing_Airline',

    # عندنا FlightStatus بدلها
    'Cancelled', 'Diverted',

    # عندنا MonthName و DayName بدلها
    'DepTimeBlk', 'ArrTimeBlk',

    # عندنا DistanceGroup بدلها
    'DivAirportLandings',
]

print(f"Before: {df.shape}")
df = df.drop(columns=cols_to_drop)
print(f"After: {df.shape}")
print(f"Memory: {df.memory_usage(deep=True).sum() / 1e9:.2f} GB")

Before: (4078318, 75)
After: (4078318, 50)
Memory: 2.34 GB


In [19]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4078318 entries, 0 to 4078317
Data columns (total 50 columns):
 #   Column                           Dtype         
---  ------                           -----         
 0   FlightDate                       datetime64[ns]
 1   Airline                          category      
 2   Origin                           category      
 3   Dest                             category      
 4   CRSDepTime                       int64         
 5   DepTime                          int64         
 6   DepDelayMinutes                  float64       
 7   DepDelay                         float64       
 8   ArrTime                          int64         
 9   ArrDelayMinutes                  float64       
 10  AirTime                          float64       
 11  CRSElapsedTime                   float64       
 12  ActualElapsedTime                float64       
 13  Distance                         float64       
 14  IATA_Code_Marketing_Airline      c

## 8. Saving Clean Data
- **Full dataset** saved as `.csv` for Power BI and team use
- **Excel sample** of 500K rows for Excel analysis  
- **Parquet file** recommended for future Python work (faster load + preserves dtypes)

> Final clean dataset: **4,078,318 rows × 50 columns — ~2.34 GB**

In [ ]:
csv_file_path = "Skyscope_Analytics_Cleaned_Full2.csv"
df.to_csv(csv_file_path, index=False)
print(f"Saved succefully {csv_file_path}" )

## A report by excell

In [20]:
big_airlines = df.groupby('Airline', observed=True).size()
big_airlines = big_airlines[big_airlines > 50000].index

# 2. report
airline_report = (df[df['Airline'].isin(big_airlines)]
                  .groupby('Airline', observed=True)
                  .agg(
                      Total_Flights=('FlightStatus', 'count'),
                      Completed=('FlightStatus', lambda x: (x == 'Completed').sum()),
                      Cancelled=('FlightStatus', lambda x: (x == 'Cancelled').sum()),
                      Diverted=('FlightStatus', lambda x: (x == 'Diverted').sum()),
                      Delayed=('IsDelayed', lambda x: (x == 1).sum()),
                      Avg_Arr_Delay=('ArrDelay', lambda x: x[x > 0].mean()),
                      Avg_Dep_Delay=('DepDelay', lambda x: x[x > 0].mean()),
                  )
                  .reset_index())

airline_report['OnTime_Rate'] = (airline_report['Completed'] - airline_report['Delayed']) / airline_report['Total_Flights'] * 100
airline_report['Cancel_Rate'] = airline_report['Cancelled'] / airline_report['Total_Flights'] * 100
airline_report['Delay_Rate']  = airline_report['Delayed']   / airline_report['Total_Flights'] * 100
airline_report['Score']       = (
    airline_report['OnTime_Rate'] * 0.5 +
    (100 - airline_report['Cancel_Rate']) * 0.3 +
    (100 - airline_report['Delay_Rate']) * 0.2
).round(1)

airline_report = airline_report.sort_values('Score', ascending=False)
print(airline_report[['Airline', 'Total_Flights', 'OnTime_Rate', 'Cancel_Rate', 'Score']].to_string(index=False))

                    Airline  Total_Flights  OnTime_Rate  Cancel_Rate  Score
                Horizon Air          56320    82.567472     1.740057   87.7
Capital Cargo International          52827    82.179567     2.803491   87.3
      SkyWest Airlines Inc.         440807    81.438135     2.106818   86.9
                  Envoy Air         150758    81.443771     2.580958   86.8
       Delta Air Lines Inc.         512114    81.394377     2.019277   86.8
          Endeavor Air Inc.         142624    80.384788     4.079958   85.9
       Alaska Airlines Inc.         129510    78.762258     2.886264   84.9
      United Air Lines Inc.         352009    77.938348     2.293691   84.4
         Mesa Airlines Inc.          71972    77.669094     4.107153   84.0
                Comair Inc.         135884    75.273763     3.937918   82.4
     American Airlines Inc.         495953    74.875442     3.777777   82.1
     Southwest Airlines Co.         731925    73.964409     2.481675   81.6
          Re

In [ ]:
airline_report.to_excel('airline_report.xlsx', index=False)